File: 0_1.baekjoon/brute-force/java/src/BOJ6987.java
import java.io.*;
import java.util.*;

public class BOJ6987 {
    private static int[][] matches;
    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        matches = new int[15][2];
        int idx = 0;
        for (int i=0;i<5;i++) {
            for (int j=i+1;j<6;j++) {
                matches[idx][0] = i;
                matches[idx++][1] = j;
            }
        }

        for (int i=0;i<4;i++) {
            int[][] scores = new int[6][3];
            StringTokenizer st = new StringTokenizer(br.readLine());
            int total = 0;
            for (int k=0;k<6;k++) {
                for (int j=0;j<3;j++) {
                    scores[k][j] = Integer.parseInt(st.nextToken());
                    total += scores[k][j];
                }
            }

            if (total == 30 && dfs(scores, 0)) {
                sb.append(1).append(" ");
            } else {
                sb.append(0).append(" ");
            }

        }
        System.out.println(sb);
    }

    private static boolean dfs(int[][] scores, int round) {
        if (round == 15) {
            return true;
        }

        int team1 = matches[round][0];
        int team2 = matches[round][1];

        if (scores[team1][0] > 0 && scores[team2][2] > 0) {
            scores[team1][0]--;
            scores[team2][2]--;
            if (dfs(scores, round + 1)) {
                return true;
            }
            scores[team1][0]++;
            scores[team2][2]++;
        }

        if (scores[team1][2] > 0 && scores[team2][0] > 0) {
            scores[team1][2]--;
            scores[team2][0]--;
            if (dfs(scores, round + 1)) {
                return true;
            }
            scores[team1][2]++;
            scores[team2][0]++;
        }

        if (scores[team1][1] > 0 && scores[team2][1] > 0) {
            scores[team1][1]--;
            scores[team2][1]--;
            if (dfs(scores, round + 1)) {
                return true;
            }
            scores[team1][1]++;
            scores[team2][1]++;
        }

        return false;
    }

}

File: 0_1.baekjoon/brute-force/kotlin/src/BOJ6987.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    repeat(4) {
        val st = StringTokenizer(readLine())
        val scores = Array(6){IntArray(3) { 0} }
        var total = 0

        for (i in 0 until 6) {
            for (j in 0 until 3) {
                scores[i][j] = st.nextToken().toInt()
                total += scores[i][j]
            }
        }

        val matches = Array(15) { IntArray(2)}
        var idx = 0
        for (i in 0 until 5) {
            for (j in i + 1 until 6) {
                matches[idx][0] = i
                matches[idx++][1] = j
            }
        }
        val sb = StringBuilder()
        if (total == 30 && dfs(scores, matches, 0)) {
            sb.append(1).append(" ")
        } else {
            sb.append(0).append(" ")
        }

        print(sb)
    }
}

private fun dfs(scores: Array<IntArray>, matches: Array<IntArray>, round: Int): Boolean {
    if (round == 15) {
        return true
    }

    val(team1, team2) = matches[round]

    if (scores[team1][0] > 0 && scores[team2][2] > 0) {
        scores[team1][0]--
        scores[team2][2]--
        if (dfs(scores, matches, round + 1)) return true
        scores[team1][0]++
        scores[team2][2]++
    }

    if (scores[team1][2] > 0 && scores[team2][0] > 0) {
        scores[team1][2]--
        scores[team2][0]--
        if (dfs(scores, matches, round + 1)) return true
        scores[team1][2]++
        scores[team2][0]++
    }

    if (scores[team1][1] > 0 && scores[team2][1] > 0) {
        scores[team1][1]--
        scores[team2][1]--
        if (dfs(scores, matches, round + 1)) return true
        scores[team1][1]++
        scores[team2][1]++
    }
    return false
}